stdlib/sys/process.gr

```diff
@@ -61,19 +78,130 @@ enum Signal {
 
 // Access command line arguments
 // @returns Array<String> The positional string arguments to the process
-import foreign wasm argv : () -> Array<String> from "stdlib-external/sys"
+export let argv = () => {
+  let argcPtr = Memory.malloc(8n)
+  let argvBufSizePtr = argcPtr + 4n
+
+  let mut err = Wasi.args_sizes_get(argcPtr, argvBufSizePtr)
+  if (err != Wasi._ESUCCESS) {
+    Memory.free(argcPtr)
+    throwError(Errors._GRAIN_ERR_SYSTEM, WasmI32.fromGrain(tagSimpleNumber(err)), 0n)
+  }
+
+  let argc = WasmI32.load(argcPtr, 0n)
+  let argvBufSize = WasmI32.load(argvBufSizePtr, 0n)
+
+  let argvPtr = Memory.malloc(argc * 4n)
+  let argvBufPtr = Memory.malloc(argvBufSize)
+
+  err = Wasi.args_get(argvPtr, argvBufPtr)
+  if (err != Wasi._ESUCCESS) {
+    Memory.free(argcPtr)
+    Memory.free(argvPtr)
+    Memory.free(argvBufPtr)
+    throwError(Errors._GRAIN_ERR_SYSTEM, WasmI32.fromGrain(tagSimpleNumber(err)), 0n)
+  }
+
+  let arr = allocateArray(argc)
+
+  let argsLength = argc * 4n
+  for (let mut i = 0n; i < argsLength; i += 4n) {
+    let strPtr = WasmI32.load(argvPtr + i, 0n)
+    let mut strLength = 0n
+    while (WasmI32.load8U(strPtr + strLength, 0n) != 0n) {
+      strLength += 1n
+    }
+
+    let grainStrPtr = allocateString(strLength)
+    Memory.copy(grainStrPtr + 8n, strPtr, strLength)
+
+    WasmI32.store(arr + i, grainStrPtr, 8n)
+  }
+
+  Memory.free(argcPtr)
+  Memory.free(argvPtr)
+  Memory.free(argvBufPtr)
+
+  WasmI32.toGrain(arr): Array<String>
+}
 
 // Access environment variables
 // @returns Array<String> The environment variables supplied to the process
-import foreign wasm env : () -> Array<String> from "stdlib-external/sys"
+export let env = () => {
+  let envcPtr = Memory.malloc(8n)
+  let envvBufSizePtr = envcPtr + 4n
+
+  let mut err = Wasi.environ_sizes_get(envcPtr, envvBufSizePtr)
+  if (err != Wasi._ESUCCESS) {
+    Memory.free(envcPtr)
+    throwError(Errors._GRAIN_ERR_SYSTEM, WasmI32.fromGrain(tagSimpleNumber(err)), 0n)
+  }
+
+  let envc = WasmI32.load(envcPtr, 0n)
+  let envvBufSize = WasmI32.load(envvBufSizePtr, 0n)
+
+  let envvPtr = Memory.malloc(envc * 4n)
+  let envvBufPtr = Memory.malloc(envvBufSize)
+
+  err = Wasi.environ_get(envvPtr, envvBufPtr)
+  if (err != Wasi._ESUCCESS) {
+    Memory.free(envcPtr)
+    Memory.free(envvPtr)
+    Memory.free(envvBufPtr)
+    throwError(Errors._GRAIN_ERR_SYSTEM, WasmI32.fromGrain(tagSimpleNumber(err)), 0n)
+  }
+
+  let arr = allocateArray(envc)
+
+  let envsLength = envc * 4n
+  for (let mut i = 0n; i < envsLength; i += 4n) {
+    let strPtr = WasmI32.load(envvPtr + i, 0n)
+    let mut strLength = 0n
+    while (WasmI32.load8U(strPtr + strLength, 0n) != 0n) {
+      strLength += 1n
+    }
+
+    let grainStrPtr = allocateString(strLength)
+    Memory.copy(grainStrPtr + 8n, strPtr, strLength)
+
+    WasmI32.store(arr + i, grainStrPtr, 8n)
+  }
+
+  Memory.free(envcPtr)
+  Memory.free(envvPtr)
+  Memory.free(envvBufPtr)
+
+  WasmI32.toGrain(arr): Array<String>
```

Comment:
Same as above.

Reply:
This one could probably be a list, yeah. Though honestly it's probably far more useful as a hash. That can probably happen in the higher level process module if we ever make it, though.